### ISSUE TITLE
Create a TypeScript backend

### ISSUE THREAD

**Issue by AttilaMihaly** [finos/morphir-elm#503]
Create a TypeScript backend for Morphir. In the first phase we only want to support:

- [ ] Generate TypeScript type definitions based on Morphir type definitions.
- [ ] Generate TypeScript code that reads data from JSON into the generated type definitions.

**Comment by ssssam** [comment]
List of Elm types that we need to consider:

 * `Bool`
 * `Int`
 * `Float`
 * `Char`
 * `String`
 * `Maybe a`
 * `List a`
 * `Dict`
 * `Set`
 * `Result`
 * tuple
 * record
 * custom type aka tagged union
 * `Order`
 * `Never`

 See also: [JSON Mapping](https://github.com/finos/morphir-elm/blob/master/docs/json-mapping.md)

**Comment by ssssam** [comment]
## Corner case #1: empty interfaces

Example from Morphir.IR.ts:

```
interface Private {}

interface Public {}

type Access = Private | Public
```

This is a straight conversion from [Morphir.IR.AccessControlled](https://package.elm-lang.org/packages/finos/morphir-elm/latest/Morphir-IR-AccessControlled) and compiles without issue.

Empty interfaces don't seem to be idiomatic TypeScript, though. TSLint has a [rule to prohibit them](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-empty-interface.md). A more idiomatic TypeScript representation might use [enums](https://www.typescriptlang.org/docs/handbook/enums.html), like this:

```
enum Access {
    Private,
    Public,
}
```

A TypeScript enum can represent a `Morphir.IR.CustomTypeDefinition` iff all variations have no arguments. So we would still need an empty interface in some cases.

**Comment by ssssam** [comment]
> ## Corner case #1: empty interfaces
> 
> Example from Morphir.IR.ts:
> 
> ```
> interface Private {}
> 
> interface Public {}
> 
> type Access = Private | Public
> ```

This example is wrong, it should be:

```
interface Private {
  kind: "private"
}

interface Public {
  kind: "public"
}

type Access = Private | Public
```

So there is never an empty interface. It could still be considered "un-idiomatic", but I think that's fine.

**Comment by AttilaMihaly** [comment]
I agree that mapping custom types with all no-arg constructors to enums.

**Comment by Cynical-Optimist** [comment]
> I agree that mapping custom types with all no-arg constructors to enums is more idiomatic.

I feel like making our code consistent within itself is probably more important than living up to TypeScript idioms though.

Consider these two custom Elm types:

```
type MyFirstType
    = KindOneAlpha
    | KindOneBeta

type MySecondType
    = KindTwoAlpha
    | KindTwoBeta
    | KindTwoGamma String Int Bool
```

KindOneAlpha and KindTwoAlpha are effectively identical constructors in the elm code. And I think that means they should be represented the same way in the TypeScript code. If the second one is going to be implemented as an interface, the first one should be too.

I think anything else would get too confusing to work with.

**Comment by Cynical-Optimist** [comment]
# Corner case #2: name conflicts

The Morphir codebase contains entities with names like `Record` and `Function`, which the backend is mapping to types and interfaces. This will create name conflicts because Typescript already has these defined as keywords.

Output from `morphir/ir/Type.ts`:
```
...
interface Function<a>{
  kind: "Function";
  arg1: a;
  arg2: any /* Unhandled type: Reference Morphir:IR.Type:type) */;
  arg3: any /* Unhandled type: Reference Morphir:IR.Type:type) */;
}

interface Record<a>{
  kind: "Record";
  arg1: a;
  arg2: Array<any /* Unhandled type: Reference Morphir:IR.Type:field) */>;
}
...
type Type<a> = ExtensibleRecord<a> | Function<a> | Record<a> | Reference<a> | Tuple<a> | Unit<a> | Variable<a>
```

According to VSCode, Both definitions create errors in the code.
